core/src/test/java/com/linecorp/armeria/server/auth/AuthorizerTest.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2018 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.server.auth;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.spotify.futures.CompletableFutures.exceptionallyCompletedFuture;
+import static java.util.concurrent.CompletableFuture.completedFuture;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import javax.annotation.Nullable;
+
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+import io.netty.channel.DefaultEventLoop;
+import io.netty.channel.EventLoop;
+
+public class AuthorizerTest {
+
+    @Nullable
+    private static EventLoop eventLoop;
+    @Nullable
+    private static ServiceRequestContext serviceCtx;
+
+    @BeforeClass
+    public static void setEventLoop() {
+        eventLoop = new DefaultEventLoop();
+        serviceCtx = mock(ServiceRequestContext.class);
+        when(serviceCtx.contextAwareEventLoop()).thenReturn(eventLoop);
```

Comment:
Though I've always been wondering whether it was really a good idea for `ServiceRequestContext` to return `EventLoop`. Mocking in unit tests would be way easier if it returned a more generic thing like `Executor` or `ScheduledExecutorService`.

Reply:
That's a good question. It is internally used often for using the desired event loop when acquiring a connection from the pool. Perhaps we can downcast there and expose `ScheduledExecutorService` there? One downside of this change would be that `ScheduledExecutorService` returns `Future` rather than `CompletableFuture`. Netty's `EventLoop` returns Netty's own `Future` so it fits better for asynchronous programming.

Reply:
Yeah I guess replacing the method is too much but adding new methods with the downcasts could make it easier to write tests. Especially, like here `Executor` is used a ton and is usually fine being `MoreExecutors.directExecutor`.